runtime/lua/vim/fs.lua

```diff
@@ -405,7 +409,29 @@ function M.normalize(path, opts)
   -- Remove trailing slashes
   path = path:gsub('(.)/$', '%1')
 
-  return path
+  -- Resolve `.` and `..` components
+  local resolved_path = ''
+  local curr_component = ''
+
+  -- Iterate through the string and process each components separated by `/`
+  -- Go one character after the end of the string to handle the last component
+  for i = 1, #path + 1 do
```

Comment:
My initial feeling  is that using `vim.split(path, '/')` might be more straightforward. Then maybe construct `resolved_path` as an array with a single `M.joinpath()` at the end?